### ISSUE TITLE
Push wiki to remote

### ISSUE THREAD

**Issue by marcrleonard** [redimp/otterwiki#24]
Is there a place in the interface the push the wiki to a git remote location? It would be cool if there was a "Push" or "Sync" feature. This could be very tricky if there are remote changes that conflict with local, but it would great to do this in the UI

**Comment by redimp** [comment]
The major problem that I see here is authentication. When pulling from a remote this might be via https and possible as an anonymous user, but when it comes to pushing you need to be authenticated in most of the cases. To implement this you would have to take care of a huge pile of problems.

**Comment by marcrleonard** [comment]
Yeah I was kind of thinking there would be an initial requirement that their credentials would have to be cached with the os. Then, presumably, the git api would be able to push. 

If a user would try to push (via a button in the ui) and it fails, you could display a helpful error and a 'set up' page outlining how to get the repo/creds set up. You could make it like how an empty GitHub repo will show you the initial set of commands to run to get up and going.

**Comment by ezcafe** [comment]
Please add this feature! I want to backup wiki to Github account

**Comment by ethereal-engineer** [comment]
I need this feature to exist. It would be easiest to achieve using public-key authentication with the git remote. But no, it's not a trivial feature. Damn.

**Comment by SnowyJaguar1034** [comment]
I too would appreciate it if such a feature existed.

**Comment by ReessKennedy** [comment]
+1

**Comment by Wrykoulaka** [comment]
+1, just tried to achieve it mapping ssh keys and adding a post-commit hook but it didnt work. I will setup a cronjob for the time being and call it a day

**Comment by tarxf** [comment]
> +1, just tried to achieve it mapping ssh keys and adding a post-commit hook but it didnt work. I will setup a cronjob for the time being and call it a day

Hey, did your cron job work? Can you post the solution?

I made the git hook work to push on save but I'm having second thoughts about the approach. The save action gets extra latency, plus the repo push is now in the Save success path. In other words, if your git provider is having issues, then you'll see an error on save.

I set it up, it's working fine, but I'll look for an asynchronous approach later. That said, it doesn't need to radically change from the way I designed it.

Let's get into how I did it, I use the **uwsgi installation** and the way I solved it is:

1. Create a repo in a git provider (I use codeberg).
2. Add that remote as an origin to the wiki repo.
3. Setup the ssh keys that the instance will use to push. This includes adding the public key in the git provider.
4. Install the post-commit hook, I am using [this one](https://xn--izc.com/posts/git-autopush/).
5. Configure the otterwiki user environment to use the ssh key to push.

This is being used in my otterwiki setup instructions, in the section [Setup Automated Push on Save](https://otterwiki.esquibits.com/How%20to%20Run%20Otterwiki%20in%20Oracle%20Cloud#setup-automated-push-on-save). The script that takes care of the setup [is here](https://codeberg.org/tarxf/otterwiki-in-oracle-cloud/src/branch/main/al9-push-on-save.sh).

**Comment by deseven** [comment]
I originally tried to do it from the container with hooks, but later discarded the idea because it turned out to be much more effort than anticipated. I ended up writing a simple script on the host that monitors the changes to the repo using inotify and then pushes them.

Script:
```bash
#!/bin/bash

REPO_PATH="/path/to/otterwiki/repository"
WATCH_PATH="$REPO_PATH/.git/refs/heads"
SSH_KEY="/path/to/the/key"
REMOTE="your_remote_name"

echo "Starting Otterwiki repo sync..."

while true; do
    inotifywait -q -e modify,create,delete,moved_to "$WATCH_PATH"

    echo "$(date): Change detected, waiting 3 seconds..."
    sleep 3

    cd "$REPO_PATH"
    GIT_SSH_COMMAND="ssh -i $SSH_KEY" git push $REMOTE --all --quiet 2>&1

    # uncomment/adjust in case you have problems with files ownership
    #chown -R 33:33 "$REPO_PATH"

    echo "$(date): Synced to $REMOTE"
done
```

Unit:
```ini
[Unit]
Description=Otterwiki repo sync
After=network.target

[Service]
Type=simple
Environment="HOME=/root"
ExecStart=/path/to/script
Restart=always
RestartSec=10

[Install]
WantedBy=multi-user.target
```

**Comment by deseven** [comment]
I made a PR that should cover this natively (#340), if someone could help with testing that would be nice. I mentioned the testing process in the second message.

**Comment by lucastucious** [comment]
For anyone wondering, I was able to make this happen with forgejo/gitea built in migration tool

![Screenshot_2026-01-15-08-56-27-69_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://github.com/user-attachments/assets/6eb9d3b7-8c62-40f1-91c2-b29b4b43b5f0)

You just need to check "make this repo a mirror" and voila

**Comment by redimp** [comment]
Hey @lucastucious, with PR #340 this feature will finally arrvie in An Otter Wiki. I plan to make a new release this week.

**Comment by deseven** [comment]
This is now a part of the [2.16.0](https://github.com/redimp/otterwiki/releases/tag/v2.16.0) release.

The features are considered to be experimental, feel free to play with them and report any issues. Note that any merging conflicts are not going to be solved automatically, keep your trees clean :)

**Comment by deseven** [comment]
There's now a guide on how to set everything up:  
https://otterwiki.com/Configuration/Repository%20Management
